From ccdf7ab09035e9817e5919af0bc626cbe823409a Mon Sep 17 00:00:00 2001
From: Isaku Yamahata <isaku.yamahata@intel.com>
Date: Mon, 13 Feb 2023 17:19:26 -0800
Subject: [PATCH 212/215] KVM: TDX: Check if all present CPUs are online

Because old TDX module requires all present CPUs to be online when
initializing TDX module(TDH.SYS.LP.INIT).  Otherwise it fails.  Because
it's difficult for users to understand the root cause, check it and
log the message.

Signed-off-by: Isaku Yamahata <isaku.yamahata@intel.com>
---
 arch/x86/kvm/vmx/tdx.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/x86/kvm/vmx/tdx.c b/arch/x86/kvm/vmx/tdx.c
index 86455b95a3e1..e53a560fa9f8 100644
--- a/arch/x86/kvm/vmx/tdx.c
+++ b/arch/x86/kvm/vmx/tdx.c
@@ -3205,6 +3205,14 @@ int __init tdx_hardware_setup(struct kvm_x86_ops *x86_ops)
 
 	/* tdx_enable() in tdx_module_setup() requires cpus lock. */
 	cpus_read_lock();
+	/*
+	 * REVERTME: The current TDX module requires TDH_SYS_LP_INIT for all
+	 * LPs to initialize.  It requires all present LPs to be online.
+	 * Once the TDX module is updated to allow offline LPs, remove this
+	 * warning.
+	 */
+	if (!cpumask_equal(cpu_online_mask, cpu_present_mask))
+		pr_warn("The old TDX module requires all present CPUs to be online to initialize.\n");
 	on_each_cpu(vmx_tdx_on, &vmx_tdx, true);	/* TDX requires vmxon. */
 	r = atomic_read(&vmx_tdx.err);
 	if (!r)
-- 
2.39.2

